_posts/henry_zhao/2019-03-06-Implement-Stardust.md

```diff
@@ -0,0 +1,287 @@
+
+[WebGL Example](https://webglsamples.org/)
+
+<image src="{{ site.urlimg }}/henry_zhao/webgl_building.gif" />
+
+There are a lot of WebGL libraries for data visualization. I will use stardust[stardust](https://stardustjs.github.io/) as an example to demonstrate how to implement it.
+
+### Stardust
+
+First install stardust:
+
+```javascript
+import Stardust from 'stardust-core';
+import 'stardust-webgl';
+```
+
+Create a platform object:
+
+```javascript
+var platform = Stardust.platform('webgl-2d', canvasNode, width, height);
+```
+
+The platform here is the based object of the canvas. All the graphics will be input into this object and drawn on a canvas. The first variable 'webgl-2d' import stardust-webgl into the platform. The second variable is a reference of html node.
+
+Start from here, we can add the graphic object like rect or circle into the canvas.
+
+Rect:
+
+```javascript
+var rectMark = Stardust.mark.rect();
+var rects = Stardust.mark.create(rectMark, platform);
+```
+
+Circle:
+
+```javascript
+var circleMark = Stardust.mark.circle();
+var circles = Stardust.mark.create(circleMark, platform);
+```
+
+To change the color of object:
+
+```javascript
+circles.attr('color', [0, 0, 0, 0.5]);
+```
+
+First three elements in the array represent the color(black). The last 0.5 is for transparency.
+
+Add the radius of the circle:
+
+```javascript
+circles.attr('radius', 2);
+```
+
+Bind data:
+
+```javascript
+var data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
+circles.data(data);
+```
+
+Add mouse event:
+
+```javascript
+    this._canvasNode.onmousedown = e => {
+      let bounds = this._canvasNode.getBoundingClientRect();
+      var x = e.clientX - bounds.left;
+      var y = e.clientY - bounds.top;
+      var pos = { x: x, y: y };
+      window.open().document.write(JSON.stringify(pos));
+    };
+  }
+ canvasNode.onmousemove = e => {
+      let bounds = this._canvasNode.getBoundingClientRect();
+      var x = e.clientX - bounds.left;
+      var y = e.clientY - bounds.top;
+      var pos = { x: x, y: y };
+      var p = this._platform.getPickingPixel(x * 2, y * 2);
+      var circles2 = Stardust.mark.create(circleMark, this._platform);
+      var remove = document.getElementById('Div1');
+      if (remove) remove.parentNode.removeChild(remove);
+      if (p) {
+        this._platform.clear();
+        this._marks.rects.render();
+        this._marks.circles.render();
+        circles2.attr('center', d => [7 * d.x + 23, 7 * d.y + 23]);
+        circles2.attr('radius', 2);
+        circles2.attr('color', [1, 0, 0, 1]);
+        circles2.data([p[0]._data[p[1]]]);
+        circles2.render();
+        var t1 = new ToolTip(
+          this._canvasNode,
+          { x: x, y: y, w: 10, h: 10 },
+          'x:' + x + ' y:' + y,
+          100,
+          1000
+        );
+        t1.check(e);
+      } else {
+        circles2.data([]);
+        circles2.render();
+      }
+    };
+```
+
```

Comment:
Add a small conclusion explaining what is obtained along with a link to a codepen or similar with the actual "live"example.